segmentation/include/pcl/segmentation/impl/extract_labeled_clusters.hpp

```diff
@@ -39,45 +39,99 @@
+    }
+
+  }
+}
+
 //////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointT> void
 pcl::extractLabeledEuclideanClusters (const PointCloud<PointT> &cloud,
                                       const typename search::Search<PointT>::Ptr &tree,
                                       float tolerance,
-                                      std::vector<std::vector<PointIndices> > &labeled_clusters,
+                                      labeled_cluster_map_t &labeled_clusters,
                                       unsigned int min_pts_per_cluster,
                                       unsigned int max_pts_per_cluster,
-                                      unsigned int)
+                                      unsigned int max_labels )
 {
-  if (tree->getInputCloud ()->points.size () != cloud.points.size ())
+  assert(tree); // ensure we have a tree
+
+  if (tree->getInputCloud ()->size () != cloud.size ())
   {
-    PCL_ERROR ("[pcl::extractLabeledEuclideanClusters] Tree built for a different point cloud dataset (%lu) than the input cloud (%lu)!\n", tree->getInputCloud ()->points.size (), cloud.points.size ());
+    PCL_ERROR ("[pcl::extractLabeledEuclideanClusters] Tree built for a different point cloud dataset (%lu) than the input cloud (%lu)!\n", tree->getInputCloud ()->size (), cloud.size ());
     return;
   }
-  // Create a bool vector of processed point indices, and initialize it to false
-  std::vector<bool> processed (cloud.points.size (), false);
 
-  std::vector<int> nn_indices;
-  std::vector<float> nn_distances;
+  // Create a bool vector of processed point indices, and initialize it to false
+  std::vector<bool> processed (cloud.size (), false);
 
-  // Process all points in the indices vector
-  for (int i = 0; i < static_cast<int> (cloud.points.size ()); ++i)
+  // Process all points in the cloud
+  for ( index_t i = 0; i < static_cast<index_t>(cloud.size()); ++i)
   {
     if (processed[i])
       continue;
 
-    std::vector<int> seed_queue;
-    int sq_idx = 0;
-    seed_queue.push_back (i);
-
     processed[i] = true;
 
-    while (sq_idx < static_cast<int> (seed_queue.size ()))
+    // label for the current point
+    const auto current_label = cloud[i].label;
+
+    // iterator in labeled_clusters for the current label
+    const auto current_label_cluster_it = labeled_clusters.find( current_label );
+
+    // max labels check
+    //  if we would have to add a new label and we're already at max, no need to proceed
+    if ( current_label_cluster_it == labeled_clusters.end() && labeled_clusters.size() >= max_labels )
+      continue;
+
+    pcl::Indices seed_queue = {i};
+
+    index_t sq_idx = 0;
+    while (sq_idx < static_cast<index_t> (seed_queue.size ()))
     {
       // Search for sq_idx
-      int ret = tree->radiusSearch (seed_queue[sq_idx], tolerance, nn_indices, nn_distances, std::numeric_limits<int>::max());
+      pcl::Indices nn_indices = {};
```

Comment:
Good catch, I had assumed the compiler would optimize this but a brief googling tells me I am probably wrong. I allocate outside and `clear` the array instead (which `radiusSearch` should take care of, but just in case...)

Reply:
>  the compiler would optimize this

C++20 brings that (compiler optimize matching pair of new and delete). In any case, it'd not apply here because the allocations/deallocations aren't happening in pairs. If only there was a way to share a buffer in a loop from one iteration to another (without needing to use a PITA allocator)